### PR TITLE
[EAGLE-526] aggregation framework for history jobs

### PR DESCRIPTION
**Pull request by wujinhu**

### CODE REVIEW

**Comment by haoch** [comment]
LGTM

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationBolt.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public void prepare(Map map, TopologyContext topologyContext, OutputCollector outputCollector) {
+        AggregationConfig.getInstance(config);
+        this.collector = outputCollector;
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        Long startTime = tuple.getLongByField("startTime");
+        LOG.info("get startTime {}", startTime);
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationBolt.java]: Don't log per event.

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    public void open(Map conf, TopologyContext context, final SpoutOutputCollector collector) {
+        AggregationConfig.getInstance(config);
+        AggregationTimeManager.instance().init(AggregationConfig.get().getZkStateConfig());
+        this.collector = collector;
+    }
+
+    @Override
+    public void nextTuple() {
+        Utils.sleep(5);
+        LOG.info("start to run");
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java]: LOG.

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        declarer.declare(new Fields("startTime"));
+    }
+
+    @Override
+    public void ack(Object messageId) {
+        //remove from processStartTime, if size == 0, then update lastUpdateTime
+        Long startTime = (Long)messageId;
+        LOG.info("succeed startTime {}", startTime);
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java]: debug log

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ public void open(Map conf, TopologyContext context, final SpoutOutputCollector c
     @Override
     public void nextTuple() {
         Utils.sleep(5);
-        LOG.info("start to run");
+        LOG.debug("start to run");
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java]: if(LOG.isDebugEnabled())

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java [hunk 5]
```diff
@@ -104,7 +104,7 @@ public void declareOutputFields(OutputFieldsDeclarer declarer) {
     public void ack(Object messageId) {
         //remove from processStartTime, if size == 0, then update lastUpdateTime
         Long startTime = (Long)messageId;
-        LOG.info("succeed startTime {}", startTime);
+        LOG.debug("succeed startTime {}", startTime);
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java]: if(LOG.isDebugEnabled())

File: eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java [hunk 6]
```diff
@@ -123,7 +123,7 @@ public void ack(Object messageId) {
     @Override
     public void fail(Object messageId) {
         Long startTime = (Long)messageId;
-        LOG.info("failed startTime {}", startTime);
+        LOG.warn("failed startTime {}", startTime);
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-aggregation/src/main/java/org/apache/eagle/jpm/aggregation/storm/AggregationSpout.java]: if(LOG.isDebugEnabled())

**Comment by yonzhang** [comment]
@wujinhu can you please add description for people to understand the changes.

**Comment by wujinhu** [comment]
Please review https://github.com/apache/incubator-eagle/pull/419 first.

This is the second step of the aggregation framework. This step uses results of the job metrics by each minute that generated by the first step. 

The framework is a storm topology. The spout decides which time range of job metrics needs to be aggregated, divide the time range into several pieces(For example, each piece is one hour) and emits these pieces to aggregate bolts.
The aggregate bolts aggregate the time range piece of job metrics that the spout assigns to them.
In aggregate bolts, they fetch job metrics by eagle server api and write the aggregate results back to eagle server.

**Pull request closed without merging**